aws/src/main/java/org/apache/iceberg/aws/RESTSigV4Signer.java

```diff
@@ -0,0 +1,117 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.hc.core5.http.EntityDetails;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpRequestInterceptor;
+import org.apache.hc.core5.http.io.entity.StringEntity;
+import org.apache.hc.core5.http.protocol.HttpContext;
+import org.apache.iceberg.exceptions.RESTException;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.signer.Aws4Signer;
+import software.amazon.awssdk.auth.signer.internal.SignerConstant;
+import software.amazon.awssdk.auth.signer.params.Aws4SignerParams;
+import software.amazon.awssdk.auth.signer.params.SignerChecksumParams;
+import software.amazon.awssdk.core.checksums.Algorithm;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.regions.providers.DefaultAwsRegionProviderChain;
+
+public class RESTSigV4Signer implements HttpRequestInterceptor {
+  public static final String AWS_HTTP_SIGV4_SIGNING_REGION = "aws.http.sigv4.signing-region";
+  public static final String AWS_HTTP_SIGV4_SIGNING_NAME = "aws.http.sigv4.signing-name";
+  public static final String AWS_HTTP_SIGV4_ACCESS_KEY_ID = "aws.http.sigv4.access-key-id";
+  public static final String AWS_HTTP_SIGV4_SECRET_ACCESS_KEY = "aws.http.sigv4.secret-access-key";
+  public static final String AWS_HTTP_SIGV4_SESSION_TOKEN = "aws.http.sigv4.session-token";
+
+  private final Aws4Signer signer = Aws4Signer.create();
+  private AwsCredentialsProvider credentialsProvider;
+
+  private String signingName;
+  private Region signingRegion;
+
+  public void initialize(Map<String, String> properties) {
+    Region defaultRegion = DefaultAwsRegionProviderChain.builder().build().getRegion();
+    this.signingRegion =
+        Region.of(properties.getOrDefault(AWS_HTTP_SIGV4_SIGNING_REGION, defaultRegion.id()));
+    this.signingName = properties.getOrDefault(AWS_HTTP_SIGV4_SIGNING_NAME, "execute-api");
```

Comment:
nit: this can be `AWS_HTTP_SIGV4_SIGNING_NAME_DEFAULT`

Reply:
+1

Reply:
done